processing/src/main/java/org/apache/druid/guice/ExtensionsLoader.java

```diff
@@ -0,0 +1,328 @@
+   * getFromExtensions().
+   */
+  @SuppressWarnings("unchecked")
+  public <T> Collection<T> getFromExtensions(Class<T> serviceClass)
+  {
+    // Classes classes are loaded once upon first request. Since the class path does
+    // not change during a run, the set of extension classes cannot change once
+    // computed.
+    //
+    // In practice, it appears the only place this matters is with DruidModule:
+    // initialization gets the list of extensions, and two REST API calls later
+    // ask for the same list.
+    Collection<?> modules = extensions.computeIfAbsent(
+        serviceClass,
+        serviceC -> new ServiceLoadingFromExtensions<>(serviceC).implsToLoad
+    );
+    //noinspection unchecked
+    return (Collection<T>) modules;
+  }
+
+  public Collection<DruidModule> getModules()
+  {
+    return getFromExtensions(DruidModule.class);
+  }
+
+  /**
+   * Find all the extension files that should be loaded by druid.
+   * <p/>
+   * If user explicitly specifies druid.extensions.loadList, then it will look for those extensions under root
+   * extensions directory. If one of them is not found, druid will fail loudly.
+   * <p/>
+   * If user doesn't specify druid.extension.toLoad (or its value is empty), druid will load all the extensions
+   * under the root extensions directory.
+   *
+   * @return an array of druid extension files that will be loaded by druid process
+   */
+  public File[] getExtensionFilesToLoad()
+  {
+    final File rootExtensionsDir = new File(extensionsConfig.getDirectory());
+    if (rootExtensionsDir.exists() && !rootExtensionsDir.isDirectory()) {
+      throw new ISE("Root extensions directory [%s] is not a directory!?", rootExtensionsDir);
+    }
+    File[] extensionsToLoad;
+    final LinkedHashSet<String> toLoad = extensionsConfig.getLoadList();
+    if (toLoad == null) {
+      extensionsToLoad = rootExtensionsDir.listFiles();
+    } else {
+      int i = 0;
+      extensionsToLoad = new File[toLoad.size()];
+      for (final String extensionName : toLoad) {
+        File extensionDir = new File(extensionName);
+        if (!extensionDir.isAbsolute()) {
+          extensionDir = new File(rootExtensionsDir, extensionName);
+        }
+
+        if (!extensionDir.isDirectory()) {
+          throw new ISE(
+              "Extension [%s] specified in \"druid.extensions.loadList\" didn't exist!?",
+              extensionDir.getAbsolutePath()
+          );
+        }
+        extensionsToLoad[i++] = extensionDir;
+      }
+    }
+    return extensionsToLoad == null ? new File[]{} : extensionsToLoad;
+  }
+
+  /**
+   * @param extension The File instance of the extension we want to load
+   *
+   * @return a URLClassLoader that loads all the jars on which the extension is dependent
+   */
+  public URLClassLoader getClassLoaderForExtension(File extension, boolean useExtensionClassloaderFirst)
+  {
+    return loaders.computeIfAbsent(
```

Comment:
Good question. I admit I didn't dive into the details when doing refactoring: I kept the original code and moved it around.

Taking a closer look now, we seem to have three ways that we load extensions: one for the Druid servers, another for `HadoopTask` and a third for `CliHadoopIndexer`. The server uses the `extensionsConfig`, the `HadoopTask` and `CliHadoopIndexer` versions always pass `false`. Why? There is no comment to indicate why.

There is, however, a comment that says that we want to load Hadoop first (in `HadoopTask`), but last (in `CliHadoopIndexer`). Modules now operate in three distinct modes:

* Extension-classloader first (configurable at run time)
* Non extension-classloader first (Hadoop)
  * Hadoop extensions first (`HadoopTask`)
  * Hadoop extensions last (`CliHadoopIndexer`)

I can't imagine this is a _feature_. It seems more like we must have run into bugs, and provided these various options to work around issues. It seems impossible for a user (or me) to predict what will break by reversing classloader search order: sometime a class reference resolves to one implementation, sometimes another. And, modules that use Hadoop must know that sometimes Hadoop is available, sometimes not.

I will say that, looking at the code, we don't mix the Hadoop and server use cases: we use one classloader or the other, but we don't seem to mix the two in the same run (he says without running Hadoop and setting breakpoints.)

Since what we have seems to work, and Hadoop is not a blocker to the test and single-server goals that this PR supports, and I'm not in a position to debug Hadoop, I'm inclined to leave well-enough alone.

Or, is there something here we need to address?

Reply:
Yeah, I agree this is all weird. From memory, I added the feature in #5321 after struggling with dependency problems with certain extensions and Hadoop jobs and not having come up with a better solution. I wish I had been able to come up with a better solution at the time, because it's a goofy feature. (I don't use it anymore on any clusters I'm currently involved with.)

At any rate, that same patch #5321 introduced the bad caching code. I should have changed the key of the loader map from `File` to `Pair<File, Boolean>`. Do you mind doing that now? You'll save me a bit of time, since otherwise I'd feel compelled to do it after you file this patch. It's more correct behavior for the function, and it won't change behavior in the normal case where the server-wide value of `useExtensionClassloaderFirst` is `false`. (All other calls pass in an explicit `false`.) So I think the risk is low.

Reply:
@gianm, thanks for the explanation. Fixed per the suggestion.